src/services/identity/SitesCacheService.ts

```diff
@@ -0,0 +1,93 @@
+import _ from "lodash"
+
+import {
+  ISOMERPAGES_REPO_PAGE_COUNT,
+  GH_MAX_REPO_COUNT,
+  ISOMER_ADMIN_REPOS,
+  GITHUB_ORG_REPOS_ENDPOINT,
+} from "@root/constants"
+import { genericGitHubAxiosInstance } from "@root/services/api/AxiosInstance"
+import type { GitHubRepositoryData, RepositoryData } from "@root/types/repoInfo"
+
+// The SitesCacheService is responsible for storing information about
+// sites that is not updated live such as lastUpdated and previewImage.
+// Changes are polled at fixed intervals, stored in cache, and served
+// to avoid long load time from live querying.
+
+const REPO_DATA_REFRESH_INTERVAL = 60 * 1000 // 1 minutes
+
+export async function getAllRepoData(
+  accessToken: string | undefined
+): Promise<RepositoryData[]> {
+  // Simultaneously retrieve all isomerpages repos
+  const paramsArr = _.fill(Array(ISOMERPAGES_REPO_PAGE_COUNT), null).map(
+    (__, idx) => ({
+      per_page: GH_MAX_REPO_COUNT,
+      sort: "full_name",
+      page: idx + 1,
+    })
+  )
+
+  const allSites = await Promise.all(
+    paramsArr.map(async (params) => {
+      const {
+        data: respData,
+      }: {
+        data: GitHubRepositoryData[]
+      } = await genericGitHubAxiosInstance.get(
+        GITHUB_ORG_REPOS_ENDPOINT,
+        accessToken
+          ? {
+              headers: { Authorization: `token ${accessToken}` },
+              params,
+            }
+          : { params }
+      )
+      return respData
+        .map((gitHubRepoData) => {
+          const {
+            pushed_at: updatedAt,
+            permissions,
+            name,
+            private: isPrivate,
+          } = gitHubRepoData
+
+          return {
+            lastUpdated: updatedAt,
+            permissions,
+            repoName: name,
+            isPrivate,
+          } as RepositoryData
+        })
+        .filter((repoData) => {
+          if (!repoData || !repoData.permissions) {
+            return false
+          }
+          return (
+            repoData.permissions.push === true &&
+            !ISOMER_ADMIN_REPOS.includes(repoData.repoName)
+          )
+        })
+    })
+  )
+  return _.flatten(allSites)
+}
+
+export class SitesCacheService {
+  private repoDataCache: RepositoryData[]
+
+  constructor() {
+    this.repoDataCache = []
+    this.renewCache()
+    setInterval(() => this.renewCache(), REPO_DATA_REFRESH_INTERVAL)
```

Comment:
Am curious if we use setInterval for a multi-instance setup, each instance will keep running this right?

In our default case of 2 instances, this will be then leading to 2 x 60 x 10 = 1200 token use?

Reply:
should we make a safer interval here that is closer to our business req? 
ie, since our last updated is to the nearest date (please correct me if I am wrong about this), could we just shift to a daily interval instead?

Reply:
@harishv7 Yea each instance will be running the interval. Perhaps once every minutes is too frequent. How about once every 5 minutes instead, leading to 2 x 12 x 10 = 240 uses per hour.

Reply:
@kishore03109 I think this interval, lets call it refresh rate, is different from the business requirements of how precise the date is. Consider if the user makes an edit today, if the refresh rate is set to once per day, the user would still see the previous `lastUpdated` date until tomorrow, in which case the user will see `lastUpdated` yesterday.

Reply:
@caando A couple of questions/comments:

1. Can I check why did we decide to use a in-memory cache over writing to the DB? Then when fetching repo info, it will be present in the same call right?

2. My concern is not really the rate, but that we are making unnecessary duplicate calls from instances. The  way to do this on a distributed system is by using a "lock" and only execute the fetches when the lock is available. If the lock is not available then it means some other instance is already fetching the data and stores it to the DB.

3. Previously,I rmb that when adding back the field, we will enforce our own definition of a last-updated at - example when a site is published. Referencing IS-169:

> We can come up with our own definition of updated at. Example, if a user pushes merges a PR then we will consider the site as updated. Then we can introduce a new column in our table to track this and hence add it back to the frontend

Are we still doing this or is the scope changed?

Reply:
@caando thanks for explaining! Regarding 

> Multiple instances is a good point and I overlooked the fact that instances might not be synchronised, so depending on which instance is serving the user, he might see different lastUpdated info.

can we confirm with @NatMaeTan if that's an acceptable tradeoff?

Reply:
> Periodically fetching and storing in-memory allows us to set a consistent limit regardless of the traffic period at a relatively low cost.

The only issue I have with this is that the total is number of instances x interval frequency. While number of instance is dynamic, our token pool is not.

Consider the case where our traffic spikes. At this point we actually need more tokens to serve requests. However, as we scale, the increase in number of instances actually causes more token usage purely from this setInterval - tokens which could be otherwise used to serve the spike.

However, I believe this would be an unlikely scenario given the history of traffic spikes. @alexanderleegs @seaerchin @kishore03109 do we have such instances before where we had to auto-scale to multiple instances?

If its an unlikely scenario, we can then use a low frequency to fetch this data.

Reply:
As discussed yesterday, the maximum number of instances CMS can be scaled up to is 5. If we limit the interval to be 5mins, the maximum token uses per hour is 12x10x5=600, which is 600/50000=1.2% of our total capacity at the current moment. I think this usage is insignificant compared to user expenditure.